### PR TITLE
[ios][android] Convert SpeechModule into universal module

### LINKED ISSUE #3077: Convert SpeechModule into universal module

### PR DESCRIPTION
**Pull request by Szymon20000**
# Why

resolves https://github.com/expo/expo/issues/3077

# How

- https://github.com/expo/expo/blob/master/guides/Creating%20Universal%20Modules.md

# Test Plan

Run TextToSpeechScreen in NCL

### CODE REVIEW

File: packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechPackage.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+import expo.core.BasePackage;
+import expo.core.ExportedModule;
+
+public class SpeechPackage extends BasePackage {
+  @Override
+  public List<ExportedModule> createExportedModules(Context reactContext) {
+    return Collections.singletonList((ExportedModule) new SpeechModule(reactContext));
+  }
+}
```
Comment (sjchmiela) [inline packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechPackage.java]: Again, I think there should be a new line at the end of file

File: packages/expo-speech/ios/EXSpeechModule.podspec [hunk 2]
```diff
@@ -0,0 +1,23 @@
+require 'json'
+
+package = JSON.parse(File.read(File.join(__dir__, '..', 'package.json')))
+
+Pod::Spec.new do |s|
+  s.name           = 'EXSpeechModule'
```
Comment (sjchmiela) [inline packages/expo-speech/ios/EXSpeechModule.podspec]: Hm… Shouldn't this be `EXSpeech`?

File: packages/expo-speech/ios/EXSpeechModule.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -0,0 +1,268 @@
+// !$*UTF8*$!
```
Comment (sjchmiela) [inline packages/expo-speech/ios/EXSpeechModule.xcodeproj/project.pbxproj]: Remove `xcodeproj`

File: packages/expo-speech/ios/EXSpeechModule/EXSpeech.h [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#import <EXCore/EXExportedModule.h>
```
Comment (sjchmiela) [inline packages/expo-speech/ios/EXSpeechModule/EXSpeech.h]:
```suggestion
// Copyright 2015-present 650 Industries. All rights reserved.

#import <EXCore/EXExportedModule.h>
```

File: packages/expo/src/Expo.ts [hunk 5]
```diff
@@ -37,6 +37,7 @@ export { Print } from 'expo-print';
 export { Accelerometer, Gyroscope, Magnetometer, MagnetometerUncalibrated } from 'expo-sensors';
 import * as SMS from 'expo-sms';
 export { SMS };
+export { Speech } from 'expo-speech';
```
Comment (sjchmiela) [inline packages/expo/src/Expo.ts]: I think we would like to keep the `import *, export` notation to let dead-code eliminators do their job

File: packages/expo-speech/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+import * as Speech from './Speech/Speech';
```
Comment (sjchmiela) [inline packages/expo-speech/src/index.ts]: Remove this file and point `main` to `Speech/Speech` 🙂

Comment (sjchmiela) [reply packages/expo-speech/src/index.ts]: For explanation see discussion at https://github.com/expo/expo/pull/2920/files/7ea7e9f76e0a1f1c0b3d321493ac8d607293d049#diff-4b06aa1fcb81c39457d7f817740f39ab

File: packages/expo-speech/ios/EXSpeechModule/EXSpeech.m [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+
+- (void)stopObserving {
+}
+
+
+- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer
+  didStartSpeechUtterance:(AVSpeechUtterance *)utterance
+{
+  id<EXEventEmitterService> emiter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
```
Comment (sjchmiela) [inline packages/expo-speech/ios/EXSpeechModule/EXSpeech.m]:
```suggestion
  id<EXEventEmitterService> emitter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
```

File: packages/expo-speech/ios/EXSpeechModule/EXSpeech.m [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  if (emiter != nil) {
```
Comment (sjchmiela) [inline packages/expo-speech/ios/EXSpeechModule/EXSpeech.m]: In ObjC calling methods on `nil` is _safe_, so we don't need to `if` this, but it's ok to leave it as far as I'm concerned

File: packages/expo-speech/ios/EXSpeechModule/EXSpeech.m [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    [emiter sendEventWithName:@"Exponent.speakingStarted" body:@{ @"id": ((EXSpeechUtteranceWithId *) utterance).utteranceId }];
+  }
+}
+
+- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer
+ didCancelSpeechUtterance:(AVSpeechUtterance *)utterance
+{
+  id<EXEventEmitterService> emiter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
```
Comment (sjchmiela) [inline packages/expo-speech/ios/EXSpeechModule/EXSpeech.m]:
```suggestion
  id<EXEventEmitterService> emitter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
```

File: packages/expo-speech/ios/EXSpeechModule/EXSpeech.m [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  if (emiter != nil) {
+    [emiter sendEventWithName:@"Exponent.speakingStopped" body:@{ @"id": ((EXSpeechUtteranceWithId *) utterance).utteranceId }];
+  }
+}
+
+- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer
+ didFinishSpeechUtterance:(AVSpeechUtterance *)utterance
+{
+  id<EXEventEmitterService> emiter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
```
Comment (sjchmiela) [inline packages/expo-speech/ios/EXSpeechModule/EXSpeech.m]:
```suggestion
  id<EXEventEmitterService> emitter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
```

**Review by sjchmiela — COMMENTED** [review]
Thanks! 👏

**Review by tsapeta — CHANGES_REQUESTED** [review]
Left some comments, but looks good 💯 

Could you also rename `stop`, `pause` and `resume` methods by adding `Async` suffix to them as they return promises? The old names should still work for at least one SDK version, but calling these functions should warn that they might be removed soon.

Also, please make sure they all have `async` keyword so `UnavailabilityError` is not thrown synchronously.

**Review by tsapeta — APPROVED** [review]
Looks good to me! 💯 
After some discussions, we will be fixing some other comments in a separate PR. This one is ready to go anyway 😉

**Pull request merged**